internal/token/token_csp.go

```diff
@@ -0,0 +1,160 @@
+}
+
+func (t *CspServerToServerTokenService) CallCSP() (string, error) {
+	var oauthPath = "/csp/gateway/am/api/auth/authorize"
+	client := &http.Client{}
+
+	req, err := http.NewRequest("POST", t.CSPBaseUrl+oauthPath, strings.NewReader(url.Values{"grant_type": {"client_credentials"}}.Encode()))
+
+	if err != nil {
+		return "", err
+	}
+
+	req.Header.Add("Authorization", encodeCSPCredentials(t.CSPClientId, t.CSPClientSecret))
+	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+
+	resp, err := client.Do(req)
+
+	if err != nil {
+		return "", err
+	}
+
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
+
+	var cspResponse CSPAuthorizeResponse
+
+	err = json.Unmarshal(body, &cspResponse)
+
+	if !hasDirectIngestScope(cspResponse.Scope) {
+		log.Println("The CSP response did not find any scope matching 'aoa:directDataIngestion' which is required for Wavefront direct ingestion.")
+	}
+
+	if err != nil {
+		return "", err
+	}
+
+	t.startOrResetTicker(cspResponse)
+
+	log.Println("ACCESS TOKEN: " + cspResponse.AccessToken)
+	return cspResponse.AccessToken, nil
+}
+
+func hasDirectIngestScope(scope string) bool {
+	if len(scope) == 0 {
+		return false
+	}
+
+	for _, s := range strings.Split(scope, " ") {
+		if strings.Contains(s, "aoa:directDataIngestion") || strings.Contains(s, "aoa/*") || strings.Contains(s, "aoa:*") {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (t *CspServerToServerTokenService) startOrResetTicker(cspResponse CSPAuthorizeResponse) {
+	tickerDelay := t.getTickerDelay(cspResponse.ExpiresIn)
+
+	if t.ticker == nil {
+		t.ticker = time.NewTicker(time.Duration(tickerDelay) * time.Second)
+		t.done = make(chan bool)
+
+		// Goroutine
+		go func() {
+			log.Printf("Scheduling a goroutine to fetch fresh CSP credentials in %d seconds\n", tickerDelay)
+			for {
+				select {
+				case <-t.done:
+					return
+				case tick := <-t.ticker.C:
+					t.mutex.Lock()
+
+					log.Println("Re-fetching CSP credentials at", tick)
+					cspAccessToken, e := t.CallCSP()
+
+					if e != nil {
+						t.AccessToken = "INVALID_TOKEN"
+					} else {
+						t.AccessToken = cspAccessToken
+					}
+
+					t.mutex.Unlock()
+				}
+			}
+		}()
+	} else {
+		t.ticker.Reset(time.Duration(tickerDelay) * time.Second)
+	}
+}
+
+func (t *CspServerToServerTokenService) getTickerDelay(expiresIn int) int {
+	retVal := 0
+
+	if expiresIn < 600 {
+		retVal = expiresIn - 30
+	} else {
+		retVal = expiresIn - 180
+	}
+
+	if retVal <= 0 {
+		return t.tickerDelay
+	}
+
+	return retVal
+}
+
+func encodeCSPCredentials(CSPClientId string, CSPClientSecret string) string {
+	return "Basic " + base64.StdEncoding.EncodeToString([]byte(CSPClientId+":"+CSPClientSecret))
+}
+
+// NewCSPServerToServerTokenService returns a TokenService instance where it will call CSP with client credentials to return an access token
+func NewCSPServerToServerTokenService(CSPBaseUrl string, CSPClientId string, CSPClientSecret string) TokenService {
```

Comment:
maybe make a CSP package so we don't have to say CSP so much in type declarations, etc.